### PR TITLE
Fix/move source dir

### PR DESCRIPTION
**Pull request by jan-gerling**

### CODE REVIEW

File: data-collection/src/main/java/refactoringml/PMDatabase.java [hunk 1]
```diff
@@ -39,11 +38,14 @@ In case of (various renames), the names are replaced, e.g.
 	 */
 	public ProcessMetricTracker renameFile(String oldFileName, String newFileName, CommitMetaData commitMetaData){
 		ProcessMetricTracker pmTracker = new ProcessMetricTracker(database.getOrDefault(oldFileName, new ProcessMetricTracker(newFileName, commitMetaData)));
-		pmTracker.setFileName(newFileName);
-		//the class file was renamed or moved, but the refactoring was not detected by refactoring miner
-		if(pmTracker.getCommitCounter() != 0){
-			pmTracker.resetCounter(commitMetaData);
+		if(oldFileName.equals(newFileName)){
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/PMDatabase.java]: I don't get these `if`s here. In which cases we would call this `renameFile` method and 1) the old file doesn't exist in the database?, or 2) the old and new file names would be the same?

Moreover, the `if(database.get(newFileName) != null)`? Why do we need it?

Can you add a few comments in this method?

File: data-collection/src/main/java/refactoringml/App.java [hunk 2]
```diff
@@ -193,7 +197,7 @@ private void processCommit(RevCommit currentCommit, GitHistoryRefactoringMiner m
 			List<RefactoringCommit> allRefactoringCommits = new ArrayList<>();
 			// stores the commit meta data
 			CommitMetaData superCommitMetaData = new CommitMetaData(currentCommit, project);
-
+			List<DiffEntry> entries = calculateDiffEntries(currentCommit);
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/App.java]: Nice that you could remove the duplication!

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 3]
```diff
@@ -33,49 +29,52 @@
...
 	}
 
-	public List<RefactoringCommit> collectCommitData(RevCommit commit, CommitMetaData superCommitMetaData, List<Refactoring> refactoringsToProcess) throws IOException {
+	public List<RefactoringCommit> collectCommitData(RevCommit commit, CommitMetaData superCommitMetaData, List<Refactoring> refactoringsToProcess, List<DiffEntry> entries) {
 		List<RefactoringCommit> allRefactorings = new ArrayList<>();
 
 		try {
-			// get the map between new path -> old path
-			HashMap<String, String> filesMap = getMapWithOldAndNewFiles(repository, commit);
+			//TODO: is this obsolete? JGit misses many files, e.g. after move source dir refactorings
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java]: This is required because sometimes RefactoringMiner returns only the "new" file names. In such cases, we need the map of the old file names in order to match things (and that's why we use the git info).

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 4]
```diff
@@ -33,49 +29,52 @@
...
+						refactoredClassFile = enforceUnixPaths(filesMap.get(refactoredClassFile));
 
 					/**
 					 * Sometimes, RMiner finds refactorings in newly introduced classes. Often, as part of larger refactorings.
 					 * (See https://github.com/tsantalis/RefactoringMiner/issues/89 for a better understanding)
 					 * We can't use those, as we need a file in the previous commit to collect metrics.
 					 * Thus, we skip this refactoring.
 					 */
-					if(fileDoesNotExist(refactoredClassFile)) {
+					if(!isClassRename(refactoring) && fileDoesNotExist(refactoredClassFile)) {
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java]: Why do we need this `!isClassRename(refactoring)` check here?

Maybe removing it would be the tool more resilient. If if faces another corner case we can't identify, it will log it here and move to the next one, without having some extra behaviour we don't understand.

File: data-collection/src/test/java/integration/canaryprojects/FluentIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+		return "https://github.com/FluentLenium/FluentLenium.git";
+	}
+
+	@Test
+	public void moveSourceRefactoring(){
+		String moveSource = "85c68373dabe32334933bdf6e67091534fc1504a";
+		assertRefactoring(getRefactoringCommits(), moveSource, "Move Source Dir", 46);
+	}
+
+	/*@Test
```
Comment (mauricioaniche) [inline data-collection/src/test/java/integration/canaryprojects/FluentIntegrationTest.java]: You want this test to be commented out?

File: data-collection/src/test/java/integration/realprojects/ApacheCommonsCSVIntegrationTest.java [hunk 6]
```diff
@@ -120,7 +120,7 @@ public void stable_CSVFormat() {
 
         // in refactorings_CSVFormat, we see that there are 82 refactorings in total.
         // after this commit, there was just one more refactoring. Thus, 81 refactorings
-        Assert.assertEquals(194, stableCommitListLevel2.get(0).getProcessMetrics().refactoringsInvolved);
+        Assert.assertEquals(195, stableCommitListLevel2.get(0).getProcessMetrics().refactoringsInvolved);
```
Comment (mauricioaniche) [inline data-collection/src/test/java/integration/realprojects/ApacheCommonsCSVIntegrationTest.java]: I see that you are counting the rename you find from JGit as a refactoring. Am I right?

Are you sure you are not double counting it? Moreover, isn't it better not to count it, i.e., only counts when RefactoringMiner identifies it? RefactoringMiner uses a bit more semantic to decide whether that was a rename or not, so only looking at file renames might end up in cases we don't fully understand.

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 7]
```diff
@@ -33,49 +29,52 @@
...
 
 				//loop over all refactored classes, multiple classes can be refactored by the same refactoring, e.g. Extract Interface Refactoring
 				for (ImmutablePair<String, String> pair : refactoredFilesAndClasses(refactoring, refactoring.getInvolvedClassesBeforeRefactoring())) {
 					// get the name of the file before the refactoring
 					// if the one returned by RMiner exists in the map, we use the one in the map instead
-					String refactoredClassFile = pair.getLeft();
-					if(filesMap.containsKey(refactoredClassFile))
-						refactoredClassFile = filesMap.get(refactoredClassFile);
+					String refactoredClassFile = enforceUnixPaths(pair.getLeft());
+					//ignore the filename from jgit for move source dirs etc, because it is unreliable (#133)
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java]: So, what you are saying here is that, for the Rename refactorings things we should always trust RefactoringMiner, but for Non Rename refactorings, we trust Git.

The problem of not using the JGit one is that, if the file doesn't exist in commit - 1, we will have an exception.

Also remember that there are refactorings in RefactoringMiner that it returns the "new name" twice, instead of the "old name". That's why we always rely on the Git one. 

In which cases did you see JGit being less reliable than RMiner?

File: data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java [hunk 8]
```diff
@@ -68,24 +68,47 @@ private void collectProcessMetricsOfRefactoredCommit(CommitMetaData superCommitM
...
+			//get all renames missed by RefactoringMiner
+			if(refactoringRenames != null)
+				jGitRenames.removeAll(refactoringRenames);
+			if(jGitRenames.size() > 0){
+				log.debug("Refactoringminer missed these refactorings: " + jGitRenames + LogUtils.createErrorState(superCommitMetadata.getCommitId(), project));
+				//update the missed renames in the PM database
+				for(ImmutablePair<String, String> rename : jGitRenames){
+					log.debug("Renamed " + rename.left + " to " + rename.right + " in PMDatabase.");
+					pmDatabase.renameFile(rename.left, rename.right, superCommitMetadata);
+					pmDatabase.reportRefactoring(rename.right, superCommitMetadata);
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java]: are you sure you want to count this rename as a refactoring? I suggest we only trust on refactorings that come from RefactoringMiner, as it uses a bit more semantic, rather than simply file renames.

File: data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java [hunk 9]
```diff
@@ -22,7 +22,7 @@ protected String getRepo() {
 	@Test
 	public void refactorings() {
 		List<RefactoringCommit> refactoringCommitList = getRefactoringCommits();
-		Assert.assertEquals(6, refactoringCommitList.size());
+		Assert.assertEquals(7, refactoringCommitList.size());
```
Comment (mauricioaniche) [inline data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java]: Why do we have one more here? Did we change the repo? Or the code now captures one more refactoring because it's better?

File: data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java [hunk 10]
```diff
@@ -45,7 +45,7 @@ public void refactoringDetails(){
 	@Test
 	public void isSubclass(){
 		List<RefactoringCommit> refactoringCommitList = getRefactoringCommits();
-		Assert.assertEquals(6, refactoringCommitList.size());
+		Assert.assertEquals(7, refactoringCommitList.size());
```
Comment (mauricioaniche) [inline data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java]: Same comment here.

File: data-collection/src/test/java/integration/toyprojects/R7ToyProjectTest.java [hunk 11]
```diff
@@ -37,4 +37,9 @@ void t1() {
 
 		Assertions.assertTrue(refactorings.stream().anyMatch(x -> x.getRefactoring().equals("Extract Interface")));
 	}
+
+	@Test
```
Comment (mauricioaniche) [inline data-collection/src/test/java/integration/toyprojects/R7ToyProjectTest.java]: It's empty?

File: data-collection/src/main/java/refactoringml/util/JGitUtils.java [hunk 12]
```diff
@@ -102,19 +106,80 @@ public static String generateCommitUrl(String repositoryUrl, String commitId, bo
...
+	}
 
-			entries.stream().forEach(e -> map.put(e.getNewPath(), e.getOldPath()));
+	//all renames detected by JGit
+	// returns the filename before and after the rename
+	public static Set<ImmutablePair<String, String>> getJGitRenames(List<DiffEntry> entries){
+		Set<ImmutablePair<String, String>> renamedClassesJGit = null;
+		if(entries  != null) {
+			Set<DiffEntry> renameEntries = entries.stream()
+					.filter(diffEntry -> diffEntry.getChangeType() == DiffEntry.ChangeType.RENAME)
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/util/JGitUtils.java:138]: Yes

File: data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java [hunk 13]
```diff
@@ -68,24 +68,47 @@ private void collectProcessMetricsOfRefactoredCommit(CommitMetaData superCommitM
 		}
 	}
 
+	//update the process metrics for all renames that were missed by RefactoringMiner
+	private void processRenames(Set<ImmutablePair<String, String>> refactoringRenames, Set<ImmutablePair<String, String>> jGitRenames, CommitMetaData superCommitMetadata) {
+		//get all renames not detected by RefactoringMiner
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java:73]: It was misleading, nice catch

File: data-collection/src/test/java/integration/toyprojects/R7ToyProjectTest.java (hunk 11 above)
Comment (jan-gerling) [reply data-collection/src/test/java/integration/toyprojects/R7ToyProjectTest.java:41]: I will add this test

File: data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java (hunk 9 above)
Comment (jan-gerling) [reply data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java:25]: We capture another refactoring, we missed before.

Comment (jan-gerling) [reply data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java:25]: image

File: data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java (hunk 10 above)
Comment (jan-gerling) [reply data-collection/src/test/java/integration/toyprojects/R4ToyProjectTest.java:48]: see comment above

File: data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java (hunk 8 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java:95]: I would still count this as a refactoring, because it is exactly what the rename refactoring describes. Also, they are pretty rare.

File: data-collection/src/test/java/integration/canaryprojects/FluentIntegrationTest.java (hunk 5 above)
Comment (jan-gerling) [reply data-collection/src/test/java/integration/canaryprojects/FluentIntegrationTest.java:25]: It was not used, but I will also add it.

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java (hunk 3 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java:50]: Thanks for the explanation.

File: data-collection/src/main/java/refactoringml/PMDatabase.java (hunk 1 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/PMDatabase.java:41]: Both checks cover cases in which the rename refactoring is incorrect.
The first checks if at least some kind of rename occurred.
The second one checks if a rename onto an already existing file occurred.

Both cases are failures of our implementation, thus I changed them to throw Exceptions.

File: data-collection/src/test/java/integration/realprojects/ApacheCommonsCSVIntegrationTest.java (hunk 6 above)
Comment (jan-gerling) [reply data-collection/src/test/java/integration/realprojects/ApacheCommonsCSVIntegrationTest.java:123]: I manually validated the new cases and they are correct. We have some tests in the Toyprojects covering these cases. Therefore, I am quite confident, that we are not double counting.

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java (hunk 7 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java:67]: The problem with JGit:
It misses quite a few rename and move source dir refactorings, because it considers many files as newly introduced, even though they are just renamed files (https://github.com/tsantalis/RefactoringMiner/issues/89, #133).
I only observed these issues for these two types of refactorings and RMiner works good in these cases, thus I trust RMiner instead of JGit.

File: data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java (hunk 4 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java:77]: I added this as a saveguard, but you are right we should handle this differently.
